libevmasm/Assembly.cpp

```diff
@@ -298,6 +505,126 @@ Json::Value Assembly::assemblyJSON(map<string, unsigned> const& _sourceIndices,
 	return root;
 }
 
+std::pair<std::shared_ptr<Assembly>, std::vector<std::string>> Assembly::fromJSON(Json::Value const& _json, vector<string> const& _sourceList, int _level)
+{
+	solRequire(_json.isObject(), AssemblyImportException, "Supplied JSON is not an object.");
+	static set<string> const validMembers{".code", ".data", ".auxdata", "sourceList"};
+	for (auto const& attribute: _json.getMemberNames())
+		solRequire(validMembers.count(attribute), AssemblyImportException, "Unknown attribute '" + attribute + "'.");
+	solRequire(_json.isMember(".code"), AssemblyImportException, "Member '.code' does not exist.");
+	solRequire(_json[".code"].isArray(), AssemblyImportException, "Member '.code' is not an array.");
+	for (auto const& codeItem: _json[".code"])
+		solRequire(codeItem.isObject(), AssemblyImportException, "Item of '.code' array is not an object.");
+
+	if (_level == 0)
+	{
+		if (_json.isMember("sourceList"))
+		{
+			solRequire(_json["sourceList"].isArray(), AssemblyImportException, "Optional member 'sourceList' is not an array.");
+			for (auto const& sourceListItem: _json["sourceList"])
+				solRequire(sourceListItem.isString(), AssemblyImportException, "Item of 'sourceList' array is not of type string.");
+		}
+	} else
+		solRequire(
+			!_json.isMember("sourceList"),
+			AssemblyImportException,
+			"Member 'sourceList' is only allowed in root JSON object."
+		);
+
+	shared_ptr<Assembly> result = make_shared<Assembly>(langutil::EVMVersion(), _level == 0, "");
+	vector<string> sourceList;
+	if (_json.isMember("sourceList"))
+	{
+		solAssert(_level == 0);
+		for (auto const& it: _json["sourceList"])
+		{
+			solRequire(
+				std::find(sourceList.begin(), sourceList.end(), it.asString()) == sourceList.end(),
+				AssemblyImportException,
+				"Items in 'sourceList' array are not unique."
+			);
+			sourceList.emplace_back(it.asString());
+		}
+	} else
+		sourceList = _sourceList;
+
+	result->importAssemblyItemsFromJSON(_json[".code"], sourceList);
+	if (_json[".auxdata"])
+	{
+		solRequire(_json[".auxdata"].isString(), AssemblyImportException, "Optional member '.auxdata' is not of type string.");
+		bytes auxdata{fromHex(_json[".auxdata"].asString())};
+		solRequire(!auxdata.empty(), AssemblyImportException, "Optional member '.auxdata' is not a valid hexadecimal string.");
+		result->m_auxiliaryData = auxdata;
+	}
+
+	if (_json.isMember(".data"))
+	{
+		solRequire(_json[".data"].isObject(), AssemblyImportException, "Optional member '.data' is not an object.");
+		Json::Value const& data = _json[".data"];
+		for (Json::ValueConstIterator dataIter = data.begin(); dataIter != data.end(); dataIter++)
+		{
+			solRequire(dataIter.key().isString(), AssemblyImportException, "Key inside '.data' is not of type string.");
+			string dataItemID = dataIter.key().asString();
```

Comment:
For the record, I just talked with @aarlt about this and we concluded that the numeric value of this ``dataItemID`` should be what ends up as the next stage in the "sub paths" - and that this may not work correctly, if we just iterate over the json keys (which are probably lexicographically sorted, which will be different starting from ``10`` subassemblies).
So plan is for @aarlt to confirm this with a test case and potentially adjust the logic here accordingly.

Reply:
I think that the coverage for this was supposed to be `test/libyul/objectCompiler/manySubObjects.yul` (containing 18 subobjects) combined with testing asm from Yul export - which we now removed.

So this is no longer covered, but I'm assuming it was already tested and the logic was adjusted. I think we should restore the coverage, but could be done separately after merging the PR.

I actually think that it might be worth restoring the asm import tests for Yul code, just without the optimizer. This part was probably quick enough to still be included in CLI tests. But just having a single test covering this would be fine too.